[reland] Create torch.distributed.shard package.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #72087

We have many sharding components currently:
torch.distributed._sharded_tensor, torch.distributed._sharding_spec,
torch.distributed._sharded_optimizer and more coming.

As a result, organizing all of this under the `torch.distributed.shard`
package. For BC reasons, I'm still keeping the old packages and have them just
reference the new package.

Differential Revision: [D33904585](https://our.internmc.facebook.com/intern/diff/D33904585/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D33904585/)!

pritamdamania committed:
[reland] Create torch.distributed.shard package.

We have many sharding components currently:
torch.distributed._sharded_tensor, torch.distributed._sharding_spec,
torch.distributed._sharded_optimizer and more coming.

As a result, organizing all of this under the `torch.distributed.shard`
package. For BC reasons, I'm still keeping the old packages and have them just
reference the new package.
ghstack-source-id: 147899768

Differential Revision: [D33904585](https://our.internmc.facebook.com/intern/diff/D33904585/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D33904585/)!

[ghstack-poisoned]

fduwjj reviewed (approved):
lgtm, thanks!

pritamdamania committed:
Update on "[reland] Create torch.distributed.shard package."


We have many sharding components currently:
torch.distributed._sharded_tensor, torch.distributed._sharding_spec,
torch.distributed._sharded_optimizer and more coming.

As a result, organizing all of this under the `torch.distributed.shard`
package. For BC reasons, I'm still keeping the old packages and have them just
reference the new package.

Differential Revision: [D33904585](https://our.internmc.facebook.com/intern/diff/D33904585/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D33904585/)!

[ghstack-poisoned]

pritamdamania committed:
Update on "[reland] Create torch.distributed.shard package."


We have many sharding components currently:
torch.distributed._sharded_tensor, torch.distributed._sharding_spec,
torch.distributed._sharded_optimizer and more coming.

As a result, organizing all of this under the `torch.distributed.shard`
package. For BC reasons, I'm still keeping the old packages and have them just
reference the new package.

Differential Revision: [D33904585](https://our.internmc.facebook.com/intern/diff/D33904585/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D33904585/)!

[ghstack-poisoned]

albanD commented on docs/source/conf.py:
```diff
@@ -346,7 +346,10 @@
-exclude_patterns = []
+exclude_patterns = [
+    # Temporarily hide certain docs until they are finalized
+    'distributed_shard.rst'
```
What is this doing? If the corresponding functions are public, then the website must document them.

pritamdamania committed:
Update on "[reland] Create torch.distributed.shard package."


We have many sharding components currently:
torch.distributed._sharded_tensor, torch.distributed._sharding_spec,
torch.distributed._sharded_optimizer and more coming.

As a result, organizing all of this under the `torch.distributed.shard`
package. For BC reasons, I'm still keeping the old packages and have them just
reference the new package.

Differential Revision: [D33904585](https://our.internmc.facebook.com/intern/diff/D33904585/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D33904585/)!

[ghstack-poisoned]

pritamdamania87 commented:
Closing in favor of https://github.com/pytorch/pytorch/pull/72141
